### PR TITLE
DAOS-4614 tests: Fixing bandit check subprocess errors

### PR DESCRIPTION
**Pull request by phender**
Resolving Bandit python code check errors for subprocess method calls
with 'shell=True'.

Skip-python-bandit: false
Test-tag-hw-large: pr,hw,large soak_smoke

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

### CODE REVIEW

**Comment by dinghwah** [comment]
Looks good. 
I ran bandit check for the attached 4 files, all the high priority errors been fixed.
Still seeing low priority bandit B603 error. I will discuss with SDL team lead to bypass this.
(1)launch.py
 >> Issue: [B603:subprocess_without_shell_equals_true] subprocess call - check for execution of untrusted input.
   Severity: Low   Confidence: High
   Location: /home/dinghwah/daos/src/tests/ftest/launch2.py:221
220             print("Running: {}".format(" ".join(cmd)))
221             return subprocess.check_output(cmd, stderr=subprocess.STDOUT)
222
--------------------------------------------------
>> Issue: [B603:subprocess_without_shell_equals_true] subprocess call - check for execution of untrusted input.
   Severity: Low   Confidence: High
   Location: /home/dinghwah/daos/src/tests/ftest/launch2.py:239
238         start_time = int(time.time())
239         return_code = subprocess.call(cmd)
240         end_time = int(time.time())

(2)slum_utils.py
Test results:
>> Issue: [B603:subprocess_without_shell_equals_true] subprocess call - check for execution of untrusted input.
   Severity: Low   Confidence: High
   Location: /home/dinghwah/daos/src/tests/ftest/util/slurm_utils2.py:178
177         cmd = ["scontrol", "show", "job", handle]
178         output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
179         match = re.search(r"JobState=([a-zA-Z]+)", str(output))

**Comment by dinghwah** [comment]
I just checked with David that we should be able to skip the bandit B603 (low priority) test.
We do not need to update the line with #nosec, since we also need to check the other high or medium priority tests.

**Comment by dinghwah** [comment]
I manually checked all files with bandit, everything OK except:
(1)Please remove #nosec from the following files:
launch.py
metrics.py   
mpio_utils.py
server_utils.py
slurm_utils.py
(2)for agent_utils.py line 211, please update full path /usr/bin/ssh 
211             ["/usr/bin/ssh", client, "-o ConnectTimeout=10",

File: src/tests/ftest/soak/soak.py [hunk 1]
```diff
@@ -222,22 +226,37 @@ def get_remote_logs(self):
         # copy the files from the remote
         # TO-DO: change scp
         this_host = socket.gethostname()
-        rsync_str = "rsync -avtr --min-size=1B"
+        cmd = [
+            "bash", "-c",
```
Comment (phender) [inline src/tests/ftest/soak/soak.py:230]: Do we need full paths for the `bash` and `rsync` commands?  I don't see Bandit complaining about it in the TestResults.

File: src/tests/ftest/soak/soak.py [hunk 2]
```diff
@@ -222,22 +226,37 @@ def get_remote_logs(self):
...
-                rsync_str, self.test_log_dir, this_host),
-            self.srun_params)
+            NodeSet.fromlist(self.hostlist_clients), cmd, self.srun_params)
         if result.exit_status == 0:
-            cmd = "cp -R -p {0}/ \'{1}\'; rm -rf {0}/*".format(
-                self.test_log_dir, self.outputsoakdir)
-            try:
-                result = process.run(cmd, shell=True, timeout=30)
-            except process.CmdError as error:
+            rm_cmd = ["rm", "-rf", "{}/*".format(self.test_log_dir)]
```
Comment (phender) [inline src/tests/ftest/soak/soak.py:237]: Do we need a full path for the `rm` command?

File: src/tests/ftest/soak/soak.py [hunk 3]
```diff
@@ -222,22 +226,37 @@ def get_remote_logs(self):
...
+            result = slurm_utils.srun(
+                NodeSet.fromlist(self.hostlist_clients), rm_cmd,
+                self.srun_params)
+            if result.exit_status == 0:
+                cp_cmd = [
+                    "cp", "-R", "-p", "{}/".format(self.test_log_dir),
```
Comment (phender) [inline src/tests/ftest/soak/soak.py:243]: Do we need a full path for the `cp` command?

File: src/tests/ftest/soak/soak.py [hunk 4]
```diff
@@ -784,8 +806,8 @@ def execute_jobs(self, jobs, pools):
         self.test_log_dir = self.log_dir + "/pass" + str(self.loop)
         self.local_pass_dir = self.outputsoakdir + "/pass" + str(self.loop)
         result = slurm_utils.srun(
-            NodeSet.fromlist(self.hostlist_clients), "mkdir -p {}".format(
-                self.test_log_dir), self.srun_params)
+            NodeSet.fromlist(self.hostlist_clients),
+            ["mkdir", "-p", self.test_log_dir], self.srun_params)
```
Comment (phender) [inline src/tests/ftest/soak/soak.py:810]: Do we need a full path for the `mkdir` command?

File: src/tests/ftest/soak/soak.py [hunk 5]
```diff
@@ -872,16 +894,15 @@ def run_soak(self, test_param):
         self.all_failed_jobs = []
         # cleanup soak log directories before test on all nodes
         result = slurm_utils.srun(
-            NodeSet.fromlist(self.hostlist_clients), "rm -rf {}".format(
-                self.log_dir), self.srun_params)
+            NodeSet.fromlist(self.hostlist_clients),
+            ["rm", "-rf", self.log_dir], self.srun_params)
```
Comment (phender) [inline src/tests/ftest/soak/soak.py:898]: Do we need a full path for the `rm` command?

File: src/tests/ftest/soak/soak.py [hunk 6]
```diff
@@ -872,16 +894,15 @@ def run_soak(self, test_param):
...
         if result.exit_status > 0:
             raise SoakTestError(
                 "<<FAILED: Soak directories not removed"
                 "from clients>>: {}".format(self.hostlist_clients))
         # cleanup test_node /tmp/soak
-        cmd = "rm -rf {}".format(self.log_dir)
         try:
-            result = process.run(cmd, shell=True, timeout=30)
+            result = process.run(["rm", "-rf", self.log_dir], timeout=30)
```
Comment (phender) [inline src/tests/ftest/soak/soak.py:905]: Do we need a full path for the `rm` command?

File: src/tests/ftest/util/slurm_utils.py [hunk 7]
```diff
@@ -139,25 +139,30 @@ def write_slurm_script(path, name, output, nodecount, cmds, uniq, sbatch=None):
...
+    Args:
+        script (str): script file suitable to by run by slurm
+        logfile (str): add -o param and generate logfile
+
+    Returns:
+        str: the job ID, which is used as a handle for other functions
+
     """
+    job_id = None
+    cmd = ["sbatch"]
```
Comment (phender) [inline src/tests/ftest/util/slurm_utils.py:151]: Do we need a full path for the `sbatch` command?

File: src/tests/ftest/util/slurm_utils.py [hunk 8]
```diff
@@ -169,8 +174,8 @@ def check_slurm_job(handle):
                 one extra UNKNOWN if the handle doesn't match a known
                 slurm job.
     """
-    cmd = "scontrol show job {}".format(handle)
-    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
+    cmd = ["scontrol", "show", "job", handle]
```
Comment (phender) [inline src/tests/ftest/util/slurm_utils.py:177]: Do we need a full path for the `scontrol` command?

File: src/tests/ftest/util/slurm_utils.py [hunk 9]
```diff
@@ -244,9 +249,9 @@ def srun(nodes, cmd, srun_params=None):
         for key, value in srun_params.items():
             params_list.extend(["--{}={}".format(key, value)])
             params = " ".join(params_list)
-    cmd = "srun --nodelist={} {} {}".format(nodes, params, cmd)
+    cmd = ["srun", "--nodelist={}".format(nodes), params] + list(cmd)
```
Comment (phender) [inline src/tests/ftest/util/slurm_utils.py:252]: Do we need a full path for the `srun` command?

File: src/tests/ftest/soak/soak.py (hunk 4 above)
Comment (dinghwah) [reply src/tests/ftest/soak/soak.py:810]: Not needed for mkdir

File: src/tests/ftest/soak/soak.py (hunk 5 above)
Comment (dinghwah) [reply src/tests/ftest/soak/soak.py:898]: not needed for rm

File: src/tests/ftest/soak/soak.py (hunk 6 above)
Comment (dinghwah) [reply src/tests/ftest/soak/soak.py:905]: not needed for rm

File: src/tests/ftest/util/slurm_utils.py (hunk 7 above)
Comment (dinghwah) [reply src/tests/ftest/util/slurm_utils.py:151]: not needed.

File: src/tests/ftest/util/slurm_utils.py (hunk 8 above)
Comment (dinghwah) [reply src/tests/ftest/util/slurm_utils.py:177]: not needed.

File: src/tests/ftest/util/slurm_utils.py (hunk 9 above)
Comment (dinghwah) [reply src/tests/ftest/util/slurm_utils.py:252]: not needed

**Comment by dinghwah** [comment]
I only add the portion of full path which bandit suggested and requested.

File: src/tests/ftest/soak/soak.py (hunk 3 above)
Comment (dinghwah) [reply src/tests/ftest/soak/soak.py:243]: not needed

File: src/tests/ftest/soak/soak.py (hunk 2 above)
Comment (dinghwah) [reply src/tests/ftest/soak/soak.py:237]: not needed

File: src/tests/ftest/soak/soak.py (hunk 1 above)
Comment (dinghwah) [reply src/tests/ftest/soak/soak.py:230]: not needed.

File: src/tests/ftest/util/slurm_utils.py (hunk 9 above)
Comment (dinghwah) [reply src/tests/ftest/util/slurm_utils.py:252]: not needed

File: src/tests/ftest/util/slurm_utils.py (hunk 8 above)
Comment (dinghwah) [reply src/tests/ftest/util/slurm_utils.py:177]: not needed.

File: src/tests/ftest/util/slurm_utils.py (hunk 7 above)
Comment (dinghwah) [reply src/tests/ftest/util/slurm_utils.py:151]: not needed.

File: src/tests/ftest/soak/soak.py (hunk 6 above)
Comment (dinghwah) [reply src/tests/ftest/soak/soak.py:905]: not needed.

File: src/tests/ftest/soak/soak.py (hunk 5 above)
Comment (dinghwah) [reply src/tests/ftest/soak/soak.py:898]: not needed.

**Pull request closed without merging**